File: inttoeng.go
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this :

// Your code here along with comments explaining your approach
/*
set s as length of num
v = ""
c = 0
loop over length of number in 3s
get n1, n2, n3 digits
append to temp c3(n3, 1)
if all 3 are 0 then just increment c and continue
if n2 is 1 then use c1 with n2*10+n1
else append temp to c2(n2) and c3(n1, 0)

if c is 0 then v = temp else v = temp + denom[c+1] + v
increment c
return trimmed v

c1 returns 10 to 19 in string
c2 returns 20 to 90 in string
c3 returns 0 to 9 along with denomination
*/
package main

import (
	"fmt"
	"strconv"
	"strings"
)

var denom []string
func numberToWords(num int) string {
	denom = []string{"", "Hundred ", "Thousand ", "Million ", "Billion "}
	if num == 0 {
		return "Zero"
	}
	s := len(strconv.Itoa(num))
	v := ""
	c := 0


	for i:=0;i<s;i=i+3 {
		temp := ""
		n1 := num%10
		num = num/10
		n2 := num%10
		num = num/10
		n3 := num%10
		num = num/10
		temp += c3(n3, 1)
		if n1 ==0 && n2 ==0 && n3== 0 {
			c++
			continue
		}
		if n2 == 1 {
			temp += c1(n2*10+n1)
		} else {
			temp += c2(n2)
			temp += c3(n1, 0)
		}
		if c == 0 {
			v = temp
		} else {
			v = temp + denom[c+1] + v
		}
		c++
	}
	return strings.TrimSpace(v)
}

func c3(x int, d int) string {
	switch(x) {
	case 0: return ""
	case 1: return "One " + denom[d]
	case 2: return "Two " + denom[d]
	case 3: return "Three " + denom[d]
	case 4: return "Four " + denom[d]
	case 5: return "Five " + denom[d]
	case 6: return "Six " + denom[d]
	case 7: return "Seven " + denom[d]
	case 8: return "Eight " + denom[d]
	case 9: return "Nine " + denom[d]
	}
	return ""
}

func c1(x int) string {
	switch(x) {
	case 10: return "Ten "
	case 11: return "Eleven "
	case 12: return "Twelve "
	case 13: return "Thirteen "
	case 14: return "Fourteen "
	case 15: return "Fifteen "
	case 16: return "Sixteen "
	case 17: return "Seventeen "
	case 18: return "Eighteen "
	case 19: return "Nineteen "
	}
	return ""
}

func c2(x int) string {
	switch(x) {
	case 0: return ""
	case 2: return "Twenty "
	case 3: return "Thirty "
	case 4: return "Forty "
	case 5: return "Fifty "
	case 6: return "Sixty "
	case 7: return "Seventy "
	case 8: return "Eighty "
	case 9: return "Ninety "
	}
	return ""
}

func MainIntToEng() {
	fmt.Println(numberToWords(1234567891)) //expected One Billion Two Hundred Thirty Four Million Five Hundred Sixty Seven Thousand Eight Hundred Ninety One
}

File: int_to_eng.cpp
//TC: O(n) where n is the number of digits in the input
//SC: O(1) since there is no extra space being consumed.

class Solution {
public:
    
    
    
    
    string numberToWords(int num) {
        unordered_map<int, string> firstmap;
        firstmap[0] = "Zero";
        firstmap[1] = "One";
        firstmap[2] = "Two";
        firstmap[3] = "Three";
        firstmap[4] = "Four";
        firstmap[5] = "Five";
        firstmap[6] = "Six";
        firstmap[7] = "Seven";
        firstmap[8] = "Eight";
        firstmap[9] = "Nine";
        firstmap[10] = "Ten";
        firstmap[11] = "Eleven";
        firstmap[12] = "Twelve";
        firstmap[13] = "Thirteen";
        firstmap[14] = "Fourteen";
        firstmap[15] = "Fifteen";
        firstmap[16] = "Sixteen";
        firstmap[17] = "Seventeen";
        firstmap[18] = "Eighteen";
        firstmap[19] = "Nineteen";

        unordered_map<int, string> secmap;
        secmap[20] = "Twenty";
        secmap[30] = "Thirty";
        secmap[40] = "Forty";
        secmap[50] = "Fifty";
        secmap[60] = "Sixty";
        secmap[70] = "Seventy";
        secmap[80] = "Eighty";
        secmap[90] = "Ninety";
        
        
        int setno = 0;
        string result = "";
        
        if(num == 0)
            return "Zero";
        
        //iterate until number becomes 0
        while(num > 0){
            
            //get three numbers in every set
            int tempnum = num%1000;
            
            //for the next iteration, get the next three numbers
            num = num/1000;
            
            //if the current three numbers are 0, then just increment the set and continue
            if(tempnum == 0){
                setno += 1;
                continue;
            }
            
            //if this is the first set of 3, then dont add any suffix
            if(setno == 0){
                //cout<<"first set"<<endl;
            } else if(setno == 1){
                //cout<<"Second set"<<endl;
                //if this is the second set of three, then add thousand before evaluating what the set will be. eg, 123456 will be 123 Thousand 456 where we evaluate 123 later
                result = " Thousand " + result;
            } else if(setno == 2){
                //similarly for million and billion
                //cout<<"Third set"<<endl;
                result = " Million " + result;
            } else if(setno == 3){
                //cout<<"Fourth set"<<endl;
                result = " Billion " + result;
            }




            //each 3 digit set

            //now in each three digit set, consider each pair and then the hundreds place
            int eachpair = tempnum%100;
            int hundpos = tempnum/100;
            bool isHund = false;
            
            //if the pair is less than 20, then directly get values from the first map
            if(eachpair < 20){
                //cout<<"pair less than 20"<<endl;
                if(eachpair != 0){
                    result = firstmap[eachpair] + result;                    
                } else {
                    isHund = true;
                }
            } else{
                //if the pair is greater than 20, then isolate each value and check
                int firstpos = eachpair%10;
                int secpos = eachpair%100;
                
                if(firstpos == 0){
                    //if the units place is 0, then just look up the pair from the second map. Eg. 20, 30, 40
                    result = secmap[eachpair] + result;
                } else{
                    //pair iterator
                    //pair iterator here is to check if we are in units place or tens place
                    int pairnum = 0;                    

                    while(eachpair > 0){
                        int eachnum = eachpair%10;
                        //if we are in units place, then just simply get the units place value from first map and add to the result
                        if(pairnum == 0){
                            result = firstmap[eachnum]  + result;
                        } else{
                            //if we are in tens place then multiply the tens place value with 10 and get the new value from the second map.
                            result = secmap[eachnum*10] +  " " + result;
                        }
                        pairnum++;
                        eachpair = eachpair/10;
                    }
                    
                }
                
                
            }

            //after we are done with the pair check, then look at the hundreds position. If this is 0, then skip. If not then get the value from the first map and append to "Hundred"
            
            if(hundpos != 0){
                    result = firstmap[hundpos] + " Hundred " + result;
            }
        
                        
            //increment set number at the end of the while loop    
            setno += 1;
        }
        
        
        //remove all extra spaces in the middle
        for(int i=0; i<result.size()-1; i++){
            if(result[i] == ' ' && result[i+1] == ' '){
                //erase format: erase(position to erase from, characters to erase)
                result.erase(i+1, 1);
            }
        }
        
        //if there is a trailing space, then trim this.
        if(result[result.size()-1] == ' ')
            result.pop_back();
        
        return result;
        
    }
};